### PR TITLE
feat: Use subscription to track changes in Gallery Item

### LINKED ISSUE #4614: Use subscription to track changes in Gallery Item
isn't something like subscription for event would help here? You should receive new event once block is finalized and loaded in indexer, so 3 blocks (block varies 7s) + 5s indexer drift

_Originally posted by @yangwao in https://github.com/kodadot/nft-gallery/issues/4558#issuecomment-1372034386_

### Context

Few changes ago a [subsquid created subscriptions for GRAPHQL](https://docs.subsquid.io/develop-a-squid/graphql-api/subscriptions/)

and we now call `nftByIdMinimal` every 2 seconds to do something similar

### The task

Instead of fetching `nftByIdMinimal` every 2 seconds use subscription to do that.

### Notes

If you need to update any indexer please open an issue in the respositories:
- kodadot/rubick
- kodadot/snek

and tag @vikiival

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [ ] Feature
- [x] Refactoring

## Context

- [x] Closes #4614
- [x] Requires deployment <>

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Comment by Jarsen136** [comment]
✅ Tested: the info of gallery item will update after subscription message coming

I relist the nft and change the price from another window. https://beta.kodadot.xyz/snek/gallery/659233203-5
Then the price of nft was changed accordingly, without polling the request. https://deploy-preview-4633--koda-nuxt.netlify.app/snek/gallery/659233203-5?redesign=true
image

image

File: utils/constants.ts [hunk 1]
```diff
@@ -36,7 +40,7 @@ export const URLS = {
     nftStorage: 'https://nft-storage.kodadot.workers.dev/',
     netlify: 'https://beta.kodadot.xyz/.netlify/functions/',
     seoCard: 'https://og-image-green-seven.vercel.app/',
-    rubick: 'https://squid.subsquid.io/rubick/v/007/graphql',
+    rubick: 'https://squid.subsquid.io/rubick/v/008/graphql',
```
Comment (Jarsen136) [inline utils/constants.ts:43]: blocked by https://github.com/kodadot/nft-gallery/pull/4610

**Comment by yangwao** [comment]
pay 50 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $50 @ 24.71 USD/KSM ~ 2.023 $KSM
🧗 [Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG](https://kusama.subscan.io/account/Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)
🔗 [0xcfccc2b68a37837494f266e27320a8983bf87ea6a40ab8295bd3fbe301570149](https://kusama.subscan.io/extrinsic/0xcfccc2b68a37837494f266e27320a8983bf87ea6a40ab8295bd3fbe301570149)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Comment by yangwao** [comment]
@vikiival merge it if good

File: components/gallery/useGalleryItem.ts [hunk 2]
```diff
@@ -34,17 +37,26 @@ export const useGalleryItem = () => {
...
     queryName: urlPrefix.value === 'rmrk' ? 'nftByIdWithoutRoyalty' : 'nftById',
     variables: {
       id: params.id,
     },
+    options: {
+      fetchPolicy: 'network-only',
+    },
+  })
+  useSubscriptionGraphql({
+    query: `   nftEntities(where: {id_eq: "${params.id}"}) {
```
Comment (vikiival) [inline components/gallery/useGalleryItem.ts:50]: Why there is a query string? Please use graphqlfile instead as written in the task desc

**Review by vikiival — CHANGES_REQUESTED** [review]
Otherwise LGTM

File: components/gallery/useGalleryItem.ts (hunk 2 above)
Comment (vikiival) [reply components/gallery/useGalleryItem.ts:50]: Anyway the query is not correct, please check `nftByIdMinimal`

Comment (Jarsen136) [reply components/gallery/useGalleryItem.ts:50]: Thanks. I have updated the query.
Now the query string is composed of querys from `queries/rmrk/subsquid/nftByIdMinimal.graphql` and `queries/nftByIdMinimal.graphql`

**Comment by roiLeo** [comment]
related with #2201

File: components/gallery/useGalleryItem.ts (hunk 2 above)
Comment (vikiival) [reply components/gallery/useGalleryItem.ts:50]: Still this is not correct

You 
```graphql
nftEntities(where: {id_eq: "${params.id}"}) {
```

nftByIdMinimal
```graphql
nft: nftEntityById(id: $id) {
```

File: utils/constants.ts (hunk 1 above)
Comment (vikiival) [reply utils/constants.ts:43]: Merged

File: composables/useSubscriptionGraphql.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  const wsClient = createClient({
+    url: wsUrl,
+  })
+  wsClient.subscribe(
+    {
+      query: `
+          subscription {
+            ${query}
+          }
+          `,
```
Comment (vikiival) [inline composables/useSubscriptionGraphql.ts:34]: I do not honestly like this
It is really prone to error.

can we have it in the similar fashion as resolving queries?

**Review by vikiival — CHANGES_REQUESTED** [review]
Smol stuff

File: composables/useSubscriptionGraphql.ts (hunk 3 above)
Comment (Jarsen136) [reply composables/useSubscriptionGraphql.ts:34]: > can we have it in the similar fashion as resolving queries?

Could you pls expand more?
- Do you mean importing the specific .graphql file and converting it into string query ?  If so, I would use this package to implement it. https://www.npmjs.com/package/import-graphql-string.

**Review by vikiival — APPROVED** [review]
Let it roll 🎲

**Comment by vikiival** [comment]
pay 70 usd

**Comment by Jarsen136** [comment]
> pay 70 usd

bounty has been paid in the previous comment.

> Please fix deepsource
> 
✅ Fixed

**Pull request merged**